app/services/api/to_user_sessions_hash2.rb

```diff
@@ -0,0 +1,73 @@
+class Api::ToUserSessionsHash2
+  def initialize(form:, user:)
+    @form = form
+    @user = user
+  end
+
+  def call
+    return Failure.new(form.errors) if form.invalid?
+
+    delete_sessions(data.select(&:deleted))
+
+    Success.new(
+      {
+        upload: new_in_mobile_app,
+        download: new_in_db + outdated,
+        deleted: deleted
+      }
+    )
+  end
+
+  private
+
+  attr_reader :form, :user
+
+  def data
+    form.to_h.data
+  end
+
+  def delete_sessions(sessions)
+    user.sessions.where(uuid: sessions.map(&:uuid)).destroy_all
+  end
+
+  def deleted
+    DeletedSession.where(user: user, uuid: data.map(&:uuid)).map(&:uuid)
```

Comment:
Not sure it changes much but I maybe you could use `.pluck(:uuid)` instead of `.map` at the end